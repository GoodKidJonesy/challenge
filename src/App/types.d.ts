export type City = {
  country: string;
  lat: number;
  lon: number;
  name: string;
};

export type WeatherInfo = {
  city: string;
  temp: number;
  icon: string;
};
