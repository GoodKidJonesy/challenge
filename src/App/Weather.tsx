import React from "react";
import { City, WeatherInfo } from "./types";

const Weather = (props: { city?: string }) => {
  const [info, setInfo] = React.useState<WeatherInfo>();

  const isPending = React.useRef(false);

  const fetchCityCoordinatess = (city: string) => {
    fetch(
      `https://api.openweathermap.org/geo/1.0/direct?q=${city}&appid=4c4f0b1876954338598a7be96c66527b`
    )
      .then((res) => res.json())
      .then((data: City[]) => {
        fetchWeather(data[0]);
      })
      .catch((err) => {
        console.log(err);
        isPending.current = false;
      });
  };

  const fetchWeather = (city: City) => {
    fetch(
      `https://api.openweathermap.org/data/2.5/weather?lat=${city.lat}&lon=${city.lon}&units=metric&appid=4c4f0b1876954338598a7be96c66527b`
    )
      .then((res) => res.json())
      .then((data) => {
        setInfo({
          city: data.name,
          temp: data.main.temp,
          icon: `https://openweathermap.org/img/wn/${data.weather[0].icon}@2x.png`,
        });
      })
      .catch((err) => {
        console.log(err);
      })
      .finally(() => {
        isPending.current = false;
      });
  };

  const fetchCurrentWeather = (city: string) => {
    isPending.current = true;
    fetchCityCoordinatess(city);
  };

  React.useLayoutEffect(() => {
    if (props.city && isPending.current === false) {
      fetchCurrentWeather(props.city);
    }
  }, [props.city]);

  return info ? (
    <>
      <h1>{info.city}</h1>

      <p>{~~info.temp} Celcius</p>

      <img src={info.icon} alt="Icon" />
    </>
  ) : null;
};

export { Weather };
