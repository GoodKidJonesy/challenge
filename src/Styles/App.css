:root {
  font-size: 10px;
  font-family: "Segoe UI", Tahoma, Geneva, Verdana, sans-serif;
  --dark: #1f1f1f;
  --light: #f1f1f1;
  --teal: #1fb8b8;
  --ruby: #950e45;
}

* {
  margin: 0;
  padding: 0;
  border: none;
  font-size: 1.5rem;
  text-transform: none;
  text-decoration: none;
  box-sizing: border-box;
}

body {
  min-height: 100vh;
  display: grid;
  place-items: center;
}

.App,
.container,
#root {
  width: 100%;
  height: 100%;
  display: grid;
  place-items: center;
}

.container,
.card {
  background-color: var(--dark);
}
#light .container,
#light .card {
  background-color: var(--light);
}

.theme-toggle {
  position: absolute;
  top: 4rem;
  right: 4rem;
  cursor: pointer;
  font-size: 2rem;
  color: var(--light);
  background-color: var(--dark);
  border: 2px solid var(--light);
  padding: 1rem;
  border-radius: 1rem;
}

#light .theme-toggle {
  color: var(--dark);
  background-color: var(--light);
  border: 2px solid var(--dark);
}

.card {
  border: 2px solid var(--ruby);
  border-radius: 1rem;
  padding: 2rem;
  display: flex;
  flex-direction: column;
  align-items: center;
  min-height: 70vh;
  aspect-ratio: 3 / 5;
}

.city-input,
.show-weather-button {
  height: 5rem;
  border-radius: 1rem;
  width: 100%;
  flex-shrink: 0;
}
.city-input {
  background-color: var(--dark);
  color: var(--light);
  border: 2px solid var(--light);
  outline: none;
}
#light .city-input {
  background-color: var(--light);
  color: var(--dark);
  border: 2px solid var(--dark);
}
.city-input:focus {
  border-color: var(--teal);
}

.show-weather-button {
  margin-top: auto;
  cursor: pointer;
  border: 2px solid var(--teal);
  background-color: var(--dark);
  color: var(--teal);
  float: right;

  transition: transform 0.15s ease-in-out, border-color 0.15s ease-in-out,
    color 0.15s ease-in-out;
}

#light .show-weather-button {
  background-color: var(--light);
}

.show-weather-button:hover,
.show-weather-button:focus {
  background-color: var(--light);
  color: var(--dark);
  border-color: var(--light);
  transform: scale(1.05);
}

#light .show-weather-button:hover,
#light .show-weather-button:focus {
  background-color: var(--dark);
  color: var(--light);
  border-color: var(--dark);
}

.weather-container {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  margin-top: 2rem;
  width: 100%;
  height: 100%;
}

.city-input {
  padding-inline: 2rem;
}

.weather-icon {
  justify-self: center;
  height: 18rem;
}

.temperature {
  font-size: 4rem;
  font-weight: 700;
  margin-bottom: 1rem;
  cursor: pointer;
  color: var(--teal);
  text-decoration: underline;
  text-underline-offset: 0.5rem;
}

.feels-like {
  font-size: 2rem;
  font-weight: 700;
  margin-bottom: 1rem;
  cursor: pointer;
  color: var(--teal);
  text-underline-offset: 0.5rem;
}

.city-name {
  font-size: 3rem;
  font-weight: 700;
  margin-bottom: 1rem;
  color: var(--teal);
}

@media (min-width: 768px) {
  .card {
    min-width: 70vw;
    min-height: auto;
    aspect-ratio: 5 / 3;
  }
}

@media (min-width: 1024px) {
  .card {
    min-width: 50vw;
    min-height: auto;
    aspect-ratio: 5 / 3;
  }
  .city-name {
    font-size: 4rem;
  }
  .temperature {
    font-size: 3rem;
  }
}
