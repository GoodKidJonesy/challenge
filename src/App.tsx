import React from "react";
import { Weather } from "./Components/Weather";
import "./Styles/App.css";

const App = () => {
  const [theme, setTheme] = React.useState("dark");
  const [input, setInput] = React.useState("");
  const [city, setCity] = React.useState("");

  const buttonLabels = {
    toggleTheme: "Toggle Theme",
    showWeather: "Show Weather",
  };

  const toggleTheme = () => {
    if (theme === "light") {
      setTheme("dark");
    } else {
      setTheme("light");
    }
  };
  const updateInputValue = (event: React.ChangeEvent<HTMLInputElement>) =>
    setInput(event.target.value);

  return (
    <div className="App" id={theme}>
      <div className="container">
        <button className="theme-toggle" onClick={toggleTheme}>
          {buttonLabels.toggleTheme}
        </button>
        <div className="card">
          <input
            className="city-input"
            role="search"
            type="text"
            placeholder="Enter a city"
            value={input}
            onChange={updateInputValue}
          />

          <Weather city={city} />

          <button
            className="show-weather-button"
            onClick={() => setCity(input)}
          >
            {buttonLabels.showWeather}
          </button>
        </div>
      </div>
    </div>
  );
};

export default App;
