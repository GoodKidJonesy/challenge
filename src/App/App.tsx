import React from "react";

import { Weather } from "./Weather";

import type { FC } from "react";

import "./app.css";

export const App: FC = () => {
  const [input, setInput] = React.useState<string>("");

  const [city, setCity] = React.useState<string>("");

  const doStuff = (event: React.ChangeEvent<HTMLInputElement>) =>
    setInput(event.target.value);

  return (
    <>
      <input role="search" type="text" value={input} onChange={doStuff} />

      <button onClick={() => setCity(input)}>Show Weather</button>

      <Weather city={city} />
    </>
  );
};
