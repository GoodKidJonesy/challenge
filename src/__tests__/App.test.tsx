import { getByTestId, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "../App";

describe("App", () => {
  it("Renders Component According To Snapshot", () => {
    const { container } = render(<App />);

    expect(container).toMatchSnapshot();
  });

  it("Checks Input Functionality", () => {
    render(<App />);

    userEvent.type(screen.getByRole("search"), "Copenhagen");
    expect(screen.getByRole("search")).toHaveValue("Copenhagen");
  });

  it("Toggles Theme", () => {
    render(<App />);
    const themeButton = screen.getByText("Toggle Theme");
    const appElement = document.querySelector(".App");

    userEvent.click(themeButton);
    expect(appElement).toHaveAttribute("id", "light");

    userEvent.click(themeButton);
    expect(appElement).toHaveAttribute("id", "dark");
  });
});
