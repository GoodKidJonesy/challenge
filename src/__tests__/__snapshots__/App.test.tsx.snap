// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Renders Component According To Snapshot 1`] = `
<div>
  <div
    class="App"
    id="dark"
  >
    <div
      class="container"
    >
      <button
        class="theme-toggle"
      >
        Toggle Theme
      </button>
      <div
        class="card"
      >
        <input
          class="city-input"
          placeholder="Enter a city"
          role="search"
          type="text"
          value=""
        />
        <button
          class="show-weather-button"
        >
          Show Weather
        </button>
      </div>
    </div>
  </div>
</div>
`;
